File: yasmin_ros/test/test_service_client_state.py
# Copyright (C) 2023 Miguel Ángel González Santamarta
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.


import time
import unittest
from threading import Thread

from yasmin import set_py_loggers
from yasmin_ros.ros_clients_cache import ROSClientsCache
from yasmin_ros import ServiceState
from yasmin_ros.basic_outcomes import SUCCEED, TIMEOUT
from yasmin_ros.yasmin_node import YasminNode

import rclpy
from rclpy.node import Node
from rclpy.executors import MultiThreadedExecutor
from std_msgs.msg import String
from example_interfaces.srv import AddTwoInts


class AuxNode(Node):
    def __init__(self):
        super().__init__("test_node")

        self.service_server = self.create_service(
            AddTwoInts, "test", self.execute_service
        )

        self.pub = self.create_publisher(String, "test", 10)

    def execute_service(self, request, response):
        time.sleep(2)
        response.sum = request.a + request.b
        return response


class TestServiceClientState(unittest.TestCase):

    @classmethod
    def setUpClass(cls):
        rclpy.init()
        set_py_loggers()

        cls.aux_node = AuxNode()
        cls.executor = MultiThreadedExecutor()
        cls.executor.add_node(cls.aux_node)

        cls.spin_thread = Thread(target=cls.executor.spin)
        cls.spin_thread.start()

    @classmethod
    def tearDownClass(cls):
        ROSClientsCache.clear_all()
        YasminNode.destroy_instance()
        rclpy.shutdown()

    def test_service_client(self):

        def create_request_cb(blackboard):
            request = AddTwoInts.Request()
            request.a = 2
            request.b = 3
            return request

        state = ServiceState(AddTwoInts, "test", create_request_cb)
        self.assertEqual(SUCCEED, state())

    def test_service_client_cache(self):
        ROSClientsCache.clear_all()
        self.assertEqual(0, ROSClientsCache.get_service_clients_count())

        def create_request_cb(blackboard):
            request = AddTwoInts.Request()
            request.a = 2
            request.b = 3
            return request

        state1 = ServiceState(AddTwoInts, "test", create_request_cb)
        self.assertEqual(1, ROSClientsCache.get_service_clients_count())

        state2 = ServiceState(AddTwoInts, "test", create_request_cb)
        self.assertEqual(1, ROSClientsCache.get_service_clients_count())

        state3 = ServiceState(AddTwoInts, "test2", create_request_cb)
        self.assertEqual(2, ROSClientsCache.get_service_clients_count())

    def test_service_client_response_handler(self):

        def create_request_cb(blackboard):
            request = AddTwoInts.Request()
            request.a = 2
            request.b = 3
            return request

        def response_handler(blackboard, response):
            return "new_outcome"

        state = ServiceState(
            AddTwoInts, "test", create_request_cb, ["new_outcome"], response_handler
        )
        self.assertEqual("new_outcome", state())

    def test_service_client_retry_wait_timeout(self):
        def create_request_cb(blackboard):
            request = AddTwoInts.Request()
            request.a = 2
            request.b = 3
            return request

        retries = 3

        ## Capture the logs
        with self.assertLogs(level="WARNING") as captured:
            state = ServiceState(
                AddTwoInts,
                "test_retry",
                create_request_cb,
                maximum_retry=retries,
                wait_timeout=0.1,
            )
            self.assertEqual(TIMEOUT, state())

        ## Check that the number of WARNING logs is correct
        self.assertEqual(
            retries,
            len(captured.records),
            msg=f"Expected {retries} WARNING logs, saw {len(captured)}.\n"
            f"Captured messages: {[r.getMessage() for r in captured.records]}",
        )

    def test_service_client_retry_response_timeout(self):
        def create_request_cb(blackboard):
            request = AddTwoInts.Request()
            request.a = 2
            request.b = 3
            return request

        retries = 3

        ## Capture the logs
        with self.assertLogs(level="WARNING") as captured:
            state = ServiceState(
                AddTwoInts,
                "test",
                create_request_cb,
                maximum_retry=retries,
                response_timeout=0.1,
            )
            self.assertEqual(TIMEOUT, state())

        ## Check that the number of WARNING logs is correct
        self.assertEqual(
            retries,
            len(captured.records),
            msg=f"Expected {retries} WARNING logs, saw {len(captured)}.\n"
            f"Captured messages: {[r.getMessage() for r in captured.records]}",
        )


if __name__ == "__main__":
    unittest.main()

File: yasmin_ros/test/test_service_client_state.cpp
// Copyright (C) 2025 Miguel Ángel González Santamarta
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.

#include <chrono>
#include <gtest/gtest.h>
#include <memory>
#include <thread>

#include "example_interfaces/srv/add_two_ints.hpp"
#include "yasmin_ros/action_state.hpp"
#include "yasmin_ros/basic_outcomes.hpp"
#include "yasmin_ros/ros_clients_cache.hpp"
#include "yasmin_ros/ros_logs.hpp"
#include "yasmin_ros/service_state.hpp"
#include "yasmin_ros/yasmin_node.hpp"

using namespace yasmin_ros;
using namespace yasmin_ros::basic_outcomes;
using namespace std::chrono_literals;

class AuxNode : public rclcpp::Node {
public:
  using AddTwoInts = example_interfaces::srv::AddTwoInts;

  AuxNode() : Node("test_node") {
    // Service server
    this->service_server_ = this->create_service<AddTwoInts>(
        "test", std::bind(&AuxNode::handle_service, this, std::placeholders::_1,
                          std::placeholders::_2));
  }

private:
  rclcpp::Service<AddTwoInts>::SharedPtr service_server_;

  void handle_service(const std::shared_ptr<AddTwoInts::Request> request,
                      std::shared_ptr<AddTwoInts::Response> response) {
    std::this_thread::sleep_for(5s);
    response->sum = request->a + request->b;
  }
};

class TestServiceClientState : public ::testing::Test {
protected:
  static std::shared_ptr<AuxNode> aux_node;
  static std::shared_ptr<rclcpp::executors::MultiThreadedExecutor> executor;
  static std::thread spin_thread;

  static void SetUpTestSuite() {
    rclcpp::init(0, nullptr);
    aux_node = std::make_shared<AuxNode>();
    executor = std::make_shared<rclcpp::executors::MultiThreadedExecutor>();
    executor->add_node(aux_node);
    spin_thread = std::thread([&]() { executor->spin(); });
  }

  static void TearDownTestSuite() {
    executor->cancel();
    if (spin_thread.joinable()) {
      spin_thread.join();
    }
    aux_node.reset();
    executor.reset();
    rclcpp::shutdown();
  }
};

std::shared_ptr<AuxNode> TestServiceClientState::aux_node = nullptr;
std::shared_ptr<rclcpp::executors::MultiThreadedExecutor>
    TestServiceClientState::executor = nullptr;
std::thread TestServiceClientState::spin_thread;

TEST_F(TestServiceClientState, TestServiceClient) {
  auto blackboard = std::make_shared<yasmin::blackboard::Blackboard>();

  auto state =
      std::make_shared<ServiceState<example_interfaces::srv::AddTwoInts>>(
          "test",
          [](std::shared_ptr<yasmin::blackboard::Blackboard> blackboard) {
            auto request = std::make_shared<
                example_interfaces::srv::AddTwoInts::Request>();
            request->a = 2;
            request->b = 3;
            return request;
          },
          -1, -1, 3); // wait_timeout, response_timeout, maximum_retry

  EXPECT_EQ((*state)(blackboard), std::string(SUCCEED));
}

TEST_F(TestServiceClientState, TestServiceClientCache) {
  ROSClientsCache::clear_all();
  EXPECT_EQ(ROSClientsCache::get_service_clients_count(), 0);

  auto state1 = std::make_shared<
      ServiceState<example_interfaces::srv::AddTwoInts>>(
      "test", [](std::shared_ptr<yasmin::blackboard::Blackboard> blackboard) {
        auto request =
            std::make_shared<example_interfaces::srv::AddTwoInts::Request>();
        request->a = 2;
        request->b = 3;
        return request;
      });
  EXPECT_EQ(ROSClientsCache::get_service_clients_count(), 1);

  auto state2 = std::make_shared<
      ServiceState<example_interfaces::srv::AddTwoInts>>(
      "test", [](std::shared_ptr<yasmin::blackboard::Blackboard> blackboard) {
        auto request =
            std::make_shared<example_interfaces::srv::AddTwoInts::Request>();
        request->a = 2;
        request->b = 3;
        return request;
      });
  EXPECT_EQ(ROSClientsCache::get_service_clients_count(), 1);

  auto state3 = std::make_shared<
      ServiceState<example_interfaces::srv::AddTwoInts>>(
      "test2", [](std::shared_ptr<yasmin::blackboard::Blackboard> blackboard) {
        auto request =
            std::make_shared<example_interfaces::srv::AddTwoInts::Request>();
        request->a = 2;
        request->b = 3;
        return request;
      });
  EXPECT_EQ(ROSClientsCache::get_service_clients_count(), 2);
}

TEST_F(TestServiceClientState, TestServiceClientResponseHandler) {
  auto blackboard = std::make_shared<yasmin::blackboard::Blackboard>();

  auto state =
      std::make_shared<ServiceState<example_interfaces::srv::AddTwoInts>>(
          "test",
          [](std::shared_ptr<yasmin::blackboard::Blackboard> blackboard) {
            auto request = std::make_shared<
                example_interfaces::srv::AddTwoInts::Request>();
            request->a = 2;
            request->b = 3;
            return request;
          },
          std::set<std::string>{"new_outcome"},
          [](std::shared_ptr<yasmin::blackboard::Blackboard> blackboard,
             std::shared_ptr<example_interfaces::srv::AddTwoInts::Response>
                 response) { return "new_outcome"; });

  EXPECT_EQ((*state)(blackboard), "new_outcome");
}

TEST_F(TestServiceClientState, TestServiceClientRetryWaitTimeout) {
  auto blackboard = std::make_shared<yasmin::blackboard::Blackboard>();

  auto state =
      std::make_shared<ServiceState<example_interfaces::srv::AddTwoInts>>(
          "test_retry",
          [](std::shared_ptr<yasmin::blackboard::Blackboard> blackboard) {
            auto request = std::make_shared<
                example_interfaces::srv::AddTwoInts::Request>();
            request->a = 2;
            request->b = 3;
            return request;
          },
          1, -1, 3); // wait_timeout=1, response_timeout=-1, maximum_retry=3

  EXPECT_EQ((*state)(blackboard), std::string(TIMEOUT));
}

TEST_F(TestServiceClientState, TestServiceClientRetryResponseTimeout) {
  auto blackboard = std::make_shared<yasmin::blackboard::Blackboard>();

  auto state =
      std::make_shared<ServiceState<example_interfaces::srv::AddTwoInts>>(
          "test",
          [](std::shared_ptr<yasmin::blackboard::Blackboard> blackboard) {
            auto request = std::make_shared<
                example_interfaces::srv::AddTwoInts::Request>();
            request->a = 2;
            request->b = 3;
            return request;
          },
          -1, 1, 3); // wait_timeout=-1, response_timeout=1, maximum_retry=3

  EXPECT_EQ((*state)(blackboard), std::string(TIMEOUT));
}

int main(int argc, char **argv) {
  ::testing::InitGoogleTest(&argc, argv);
  return RUN_ALL_TESTS();
}
